src/java.base/share/classes/java/lang/reflect/code/type/CoreTypeFactory.java

```diff
@@ -90,72 +92,103 @@ public TypeElement constructType(TypeElement.ExternalizedTypeElement tree) {
      */
     public static final TypeElementFactory JAVA_TYPE_FACTORY = new TypeElementFactory() {
         @Override
-        public TypeElement constructType(TypeElement.ExternalizedTypeElement tree) {
+        public JavaType constructType(TypeElement.ExternalizedTypeElement tree) {
```

Comment:
Several changes here:
* arrays now parsed like everything else (before arrays were sort of stashed onto the side, and then re-attached later)
* type arguments are no longer parsed upfront. Each sigil will determine (a) how many type arguments are required and (b) what are the constraints on such arguments (e.g. in some cases we only want "classes" in other cases we might be ok with "everything except wildcards")
* We now check well-formedness in every branch of the code, and reject illegal combinations, such as `? extends ? extends` or `int<String>`.